File: frontend/src/services/rateTiers.js
// iter143 — Amount-tiered exchange rates (client-side twin of
// backend services/rate_tiers.py). A tier applies when the sent amount
// >= tier.min_amount; the tier with the HIGHEST matching min wins.

export function pickTierFromList(tiers, amount) {
  const list = Array.isArray(tiers) ? tiers : [];
  const amt = Number(amount);
  if (!list.length || !Number.isFinite(amt) || amt <= 0) return null;
  const sorted = [...list].sort((a, b) => Number(b.min_amount) - Number(a.min_amount));
  return sorted.find((t) => amt >= Number(t.min_amount)) || null;
}

export function pickTier(rate, amount) {
  return pickTierFromList(rate?.tiers, amount);
}

// Effective rate for the exchange preview: tier override when it matches,
// base rate_normal / rate_vip otherwise.
export function effectiveClientRate(rate, amount, isVip) {
  if (!rate) return 0;
  const tier = pickTier(rate, amount);
  const tierRate = tier ? Number(isVip ? tier.rate_vip : tier.rate_normal) : 0;
  if (tierRate > 0) return tierRate;
  return Number(isVip ? rate.rate_vip : rate.rate_normal) || 0;
}

File: backend/services/rate_tiers.py
"""Amount-tiered exchange rates — iter143.

A rate document may carry an optional `tiers` list:
    tiers: [{min_amount, rate_normal, rate_vip, real_rate?}, ...]

A tier applies when the SENT amount (amount_from) >= tier.min_amount; the
tier with the HIGHEST matching min_amount wins. When no tier matches (or
none is configured) the base rate_normal / rate_vip / real_rate on the
document apply. Mirrors the payment-account selection rule in
`services/payment_accounts.py` so account and rate always stay in sync.
"""
from typing import Any, Optional


def normalize_tiers(raw: Any) -> list[dict]:
    """Valid tier dicts sorted by min_amount DESC (highest first)."""
    if not isinstance(raw, list):
        return []
    out = []
    for t in raw:
        if not isinstance(t, dict):
            continue
        try:
            mn = float(t.get("min_amount") or 0)
        except (TypeError, ValueError):
            continue
        if mn < 0:
            continue
        out.append(t)
    out.sort(key=lambda t: float(t["min_amount"]), reverse=True)
    return out


def pick_tier(rate_doc: Optional[dict], amount: Optional[float]) -> Optional[dict]:
    if not rate_doc or amount is None:
        return None
    try:
        amt = float(amount)
    except (TypeError, ValueError):
        return None
    for tier in normalize_tiers(rate_doc.get("tiers")):
        if amt >= float(tier["min_amount"]):
            return tier
    return None


def _positive(v: Any) -> Optional[float]:
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return f if f > 0 else None


def effective_rates(rate_doc: Optional[dict], amount: Optional[float]) -> dict:
    """{rate_normal, rate_vip, real_rate} applying the matching tier (if any).
    Tier fields fall back to the base document values when absent/invalid."""
    base = rate_doc or {}
    tier = pick_tier(base, amount) or {}
    return {
        "rate_normal": _positive(tier.get("rate_normal")) or float(base.get("rate_normal") or 0.0),
        "rate_vip": _positive(tier.get("rate_vip")) or float(base.get("rate_vip") or 0.0),
        "real_rate": _positive(tier.get("real_rate")) or _positive(base.get("real_rate")),
    }
